### ISSUE TITLE
Compaction process almost caused OOM Kill.

### ISSUE THREAD

**Issue by shay-berman** [prometheus/prometheus#8357]
Compaction process almost caused OOM Kill.
Is it possible to reduce the memory usage during compaction?  what is the right compaction settings I should use?

**What did you do?**
Have prometheus with 1 million series and after only 7 days of run I saw spike of 20GB (from steady stay of 5GB).
According to the prometheus logs the spike started and finished during 4 compaction actions (15 minutes all, see logs below).

metrics `process_resident_memory_bytes` and `container_memory_working_set_bytes` during the spike.
> Looking on  `process_resident_memory_bytes` it indicates for almost OOM kill (because VM with 21GB memory limit), while on `container_memory_working_set_bytes`  no risk of OOM.  Question: which one of these metrics should I look for to indicate the real memory usage that can cause OOMKill  (We have memory.limit of 21GB on the prometheus pod)

![image](https://user-images.githubusercontent.com/1079649/104135293-0c370980-5398-11eb-876a-4edf404beff0.png)

Picture with compaction duration and process_resident_memory_bytes at the same time of the spike:
![image](https://user-images.githubusercontent.com/1079649/104154796-737fa880-53ee-11eb-8264-093148ff394a.png)

* RSS and compaction few days before the spike
Looking in the last few days you can see there is compaction that increase RSS but usually go up from 5GB to 10GB at most, but the big spike I mentioned above was from 5GB to 20GB which could lead to OOM if it goes up a bit more.
![image](https://user-images.githubusercontent.com/1079649/104154615-ffdd9b80-53ed-11eb-9848-b895e9d1ccc3.png)
![image](https://user-images.githubusercontent.com/1079649/104154753-5c40bb00-53ee-11eb-8e10-42dd5fca98b6.png)

**What did you expect to see?**
Expect compaction process to not cause OOM.

**What did you see instead? Under which circumstances?**
just spike from 5GB to 20GB during 15 minutes of compaction process.

**Environment**

* System information:

  - prometheus [v2.22.1](https://github.com/prometheus/prometheus/releases/tag/v2.22.1) (Nov 5, 2020)
  - Installed on GKE with [prometheus operator](https://github.com/prometheus-operator/prometheus-operator), installed via [helm chart](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack).
  - Running on instance with 4 cores and 26 GB RAM (eventually GKE take 2GB so eventually its 24GB. And I set prometheus container memory limit to 21GB)

   - Prometheus configuration
```
- job_name: rlec
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 30s
  metrics_path: /
  scheme: https
  tls_config:
    insecure_skip_verify: true
  static_configs:
  - targets:
     - target1
     - target2
     # And many other targets
```

  - Prometheus important Command-Line Flags
```
storage.remote.flush-deadline		1m
storage.remote.read-concurrent-limit	10
storage.remote.read-max-bytes-in-frame	1048576
storage.remote.read-sample-limit	50000000
storage.tsdb.allow-overlapping-blocks	false

storage.tsdb.max-block-duration	18d
storage.tsdb.min-block-duration	2h

storage.tsdb.no-lockfile		        true
storage.tsdb.retention			0s
storage.tsdb.retention.size		0B
storage.tsdb.retention.time		180d
storage.tsdb.wal-compression	true
storage.tsdb.wal-segment-size	0B
```
  - Note: the `storage.tsdb.max-block-duration` is [automatically](https://prometheus.io/docs/prometheus/latest/storage/#compaction) set to 18d which is 10% of the `storage.tsdb.retention.time`.

* Logs (prometheus container logs during the spike):
```
level=info ts=2021-01-09T21:00:40.863Z caller=compact.go:494 component=tsdb msg="write block" mint=1610215200000 maxt=1610222400000 ulid=01EVMF5Y1AWGX4BCM2XEAGV9BR duration=34.805323121s
level=info ts=2021-01-09T21:00:45.107Z caller=head.go:809 component=tsdb msg="Head GC completed" duration=1.227377019s
level=info ts=2021-01-09T21:00:45.460Z caller=checkpoint.go:96 component=tsdb msg="Creating checkpoint" from_segment=756 to_segment=765 mint=1610222400000
level=info ts=2021-01-09T21:01:00.318Z caller=head.go:889 component=tsdb msg="WAL checkpoint complete" first=756 last=765 duration=14.858391783s
level=info ts=2021-01-09T21:02:02.906Z caller=compact.go:440 component=tsdb msg="compact blocks" count=3 mint=1610193600000 maxt=1610215200000 ulid=01EVMF7NM2VQ20YQAV190E9Q7Z sources="[01EVKTJRF237RFYD9PMVNM1WE9 01EVM1EFQ0ATM35S7DN74T4JKF 01EVM8A6Z0Q8EHX5A8MVTY9X6Z]" duration=59.927493283s
level=info ts=2021-01-09T21:02:05.794Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVKTJRF237RFYD9PMVNM1WE9
level=info ts=2021-01-09T21:02:05.895Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVM8A6Z0Q8EHX5A8MVTY9X6Z
level=info ts=2021-01-09T21:02:05.972Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVM1EFQ0ATM35S7DN74T4JKF
level=info ts=2021-01-09T21:03:51.400Z caller=compact.go:440 component=tsdb msg="compact blocks" count=3 mint=1610150400000 maxt=1610215200000 ulid=01EVMF9NJAHMVQW64M2KYDJ548 sources="[01EVK61AMM4SXGB7KZBX1E2DJG 01EVKTMHEVY2BS2E5TD3PEV9FC 01EVMF7NM2VQ20YQAV190E9Q7Z]" duration=1m42.941761733s
level=info ts=2021-01-09T21:03:54.299Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVK61AMM4SXGB7KZBX1E2DJG
level=info ts=2021-01-09T21:03:54.602Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVKTMHEVY2BS2E5TD3PEV9FC
level=info ts=2021-01-09T21:03:54.930Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVMF7NM2VQ20YQAV190E9Q7Z
level=info ts=2021-01-09T21:07:27.878Z caller=compact.go:440 component=tsdb msg="compact blocks" count=3 mint=1610020800000 maxt=1610215200000 ulid=01EVMFD00PE0X030VB4M1JVNKF sources="[01EVGKPQ5AB8PGF53F1QPZNT5H 01EVJHG5NDSZ8QKK8QY1D0QYAT 01EVMF9NJAHMVQW64M2KYDJ548]" duration=3m30.415932572s
level=info ts=2021-01-09T21:07:31.220Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVMF9NJAHMVQW64M2KYDJ548
level=info ts=2021-01-09T21:07:31.882Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVGKPQ5AB8PGF53F1QPZNT5H
level=info ts=2021-01-09T21:07:32.879Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVJHG5NDSZ8QKK8QY1D0QYAT
level=info ts=2021-01-09T21:12:16.219Z caller=compact.go:440 component=tsdb msg="compact blocks" count=3 mint=1609632000000 maxt=1610215200000 ulid=01EVMFKMYFVKFRKTW5RZJCDJKJ sources="[01EV8WD1S7EC5ZP9ZDAPP9R2Z7 01EVEP0FWPQ284HH7ZWXQFH8CZ 01EVMFD00PE0X030VB4M1JVNKF]" duration=4m40.715703432s
level=info ts=2021-01-09T21:12:19.114Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EV8WD1S7EC5ZP9ZDAPP9R2Z7
level=info ts=2021-01-09T21:12:19.915Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVEP0FWPQ284HH7ZWXQFH8CZ
level=info ts=2021-01-09T21:12:21.129Z caller=db.go:1152 component=tsdb msg="Deleting obsolete block" block=01EVMFD00PE0X030VB4M1JVNKF
level=info ts=2021-01-09T23:00:45.013Z caller=compact.go:494 component=tsdb msg="write block" mint=1610222400000 maxt=1610229600000 ulid=01EVMP1NF1PYG5DJJV5J02FG00 duration=38.772847679s
level=info ts=2021-01-09T23:00:48.960Z caller=head.go:809 component=tsdb msg="Head GC completed" duration=1.286573607s
```

> NOTE: I will mentioned that I have few more prometheus instances that do the same but on different targets. All of them started at the same time and I saw this spike at the same time on all of them!  Here is a wider graph with more prometheus instances at the same time:
![image](https://user-images.githubusercontent.com/1079649/104151001-7d041300-53e4-11eb-8b51-33447eb61efb.png)

# references
I actually search about this and found few issues on the same thing but not sure what is the proposed solution.
1. https://github.com/prometheus/prometheus/issues/4110 but was closed with no resolution.
2. https://github.com/prometheus-operator/prometheus-operator/issues/3724 -> looks like there is no way to set `storage.tsdb.max-block-duration`
3. https://github.com/prometheus/prometheus/issues/6184 -> but again no resolution.
4. https://github.com/prometheus/prometheus/issues/7483 -> looks more related to 32bit VM, so not sure if related.

**Comment by roidelapluie** [comment]
Mmh, could you try to update Prometheus first? There were improvements in compation in 2.23. ( https://github.com/prometheus/prometheus/pull/8192 )

**Comment by roidelapluie** [comment]
cc @codesome @bwplotka do you see a bug here, or something we could improve? Thanks!

**Comment by shay-berman** [comment]
Thanks @roidelapluie  for the fast reply.
I added more info to the ticket.

**Comment by brian-brazil** [comment]
What does go_memstats_alloc_bytes show over the same period?

The working set tells us nothing, we know that compaction touches lots of data but the kernel can promptly evict that so it's not really memory usage.

**Comment by shay-berman** [comment]
Hi @brian-brazil 

1. The go_memstats_alloc_bytes was a bit increased (from 3.3GB go up to 4GB only) during this compaction time but NO spike at all on this metric. 

![image](https://user-images.githubusercontent.com/1079649/104184551-48ac4900-541c-11eb-94f1-2eae7ee732a7.png)


Another view but in different panels:
![image](https://user-images.githubusercontent.com/1079649/104182902-d2a6e280-5419-11eb-9efc-0d437323476a.png)

Another view on all the prometheus instances (not only single one) so you can see`go_memstats_alloc_bytes` just go up by 1GB no more BUT the `process_resident_memory_bytes` is spiking from 5GB to 
![image](https://user-images.githubusercontent.com/1079649/104182650-6fb54b80-5419-11eb-8dcb-018c6044e2ee.png)

2. So according to your [blog](https://www.robustperception.io/analysing-prometheus-memory-usage) alarge diff between `process_resident_memory_bytes` to `go_memstats_alloc_bytes` could indicate spiky memory usage, or fragmentation issues. Is it relevant in my case? if so what is the action item to reduce it?  

3. You mentioned `container_memory_working_set_bytes` tell us nothing. Could please explain why? according to this [blog](https://medium.com/@eng.mohamed.m.saeed/memory-working-set-vs-memory-rss-in-kubernetes-which-one-you-should-monitor-8ef77bf0acee) K8s use this metric to indicate OOMKill. And we also saw that this specific metric is quite similar to the memory utilization GKE show in its console on the pods so I assumed its the right metric to follow for any OOM indication.

4. You mentioned that `compaction touches lots of data but the kernel can promptly evict that so it's not really memory usage.` did you refer to the `process_resident_memory_bytes`?   This metric is the one that caused me to open this ticket because it spiked from 5GB to 20GB.  So if it reached 21GB does it mean it will get OOMkill or this `process_resident_memory_bytes` indicate something else.

**Comment by brian-brazil** [comment]
> go_memstats_alloc_bytes was a bit increased (from 3.3GB go up to 4GB only)

That's a little bit bigger than I'd expect from compaction, do you have massive symbol tables?

This is also the only memory that Prometheus actually needs, so this sounds like a kernel memory management issue rather than a Prometheus issue as it's OOMkilling rather than reclaiming reclaimable memory.

However you didn't actually have an OOMkill, so there's not an issue here. Only that these metrics have non-obvious meanings, and are thus not particularly useful for alerting. You unfortunately need to dig deep into both the kernel docs and cadvisor code to figure out what they actually mean.

**Comment by shay-berman** [comment]
@brian-brazil 

Not sure exactly what massive symbol tables are, but according to the metric `prometheus_tsdb_symbol_table_size_bytes`

![image](https://user-images.githubusercontent.com/1079649/104193500-da21b800-5428-11eb-8536-32c482cef54a.png)

And it looks quite stable graph if u look 5 days on this metric 
![image](https://user-images.githubusercontent.com/1079649/104193726-19500900-5429-11eb-97bb-9e3368c5092b.png)


Does the smbol table input above give us some signs about what is the problem or if there is a problem at all?
Which metric I should look for alerting before OOM of the prometheus, is it `process_resident_memory_bytes` or something else?

**Comment by brian-brazil** [comment]
> Not sure exactly what massive symbol tables are, but according to the metric prometheus_tsdb_symbol_table_size_bytes

That's tiny, nothing to worry about. I'm not sure why it's so big then, but a few hundred MB for compaction isn't exactly a problem.

> Which metric I should look for alerting before OOM of the prometheus, is it process_resident_memory_bytes or something else?

I'm not sure offhand, that it sounds like container_memory_working_set_bytes might be it but this is a Linux internal question not a Prometheus question.

**Comment by shay-berman** [comment]
Thanks @brian-brazil,   appreciate your responsiveness.

Does the `process_resident_memory_bytes` is the right metric alert about memory? for example send alert if its reaching to the VM RAM size?

Is there any way to make the compaction process to be less memory consuming to avoid OOM kill?

**Comment by brian-brazil** [comment]
Compaction already uses as little memory as it can, and the math around memory is much more complicated than that as we're only talking about OOMs here not anything performance related which is what you more care about for a binary like Prometheus that depends on page cache. I'd suggest taking this to the -users mailing list.

**Comment by shay-berman** [comment]
@brian-brazil, 
Prometheus `process_resident_memory_bytes` going from 5GB to 20GB during compaction time - is it as expected?

**Comment by sandrom** [comment]
> Compaction already uses as little memory as it can, and the math around memory is much more complicated than that as we're only talking about OOMs here not anything performance related which is what you more care about for a binary like Prometheus that depends on page cache.

I would expect to care about OOM even more than about performance, because OOM kills are going to hit performance badly, or did I misunderstand you here?

**Comment by brian-brazil** [comment]
If you're seeing OOM kills, then your performance is going to have issues well before that as you'll be trashing the page cache.

**Comment by shay-berman** [comment]
@brian-brazil I got no answer from the prometheus-mailing list so far.
Actually not sure how to address this dissonance between very low memory collected from `container_memory_working_set_bytes` vs high spike in memory from 5G-20G collected on `process_resident_memory_bytes`.

For some reason it happened during prometheus compaction process (BTW usually evet 6 hours it do the compaction but more of the time the spikes in memory are low and not risky, while once in a day or 2 it goes up to huge spike that risk for OOM).

What options do I have in such case?
- should I extend the memory of the instance to prevent OOMkill or there is a way to make the compation process with lower memory consumption?

Here is another graph of another spike:
![image](https://user-images.githubusercontent.com/1079649/104639758-74dff800-56b0-11eb-910f-009821274b32.png)

**Comment by SuperQ** [comment]
What is the OS / Linux version for this?

**Comment by shay-berman** [comment]
Hi @SuperQ 

Here is more detail per your request:
- Its run on GKE version `v1.16.15-gke.6000` and the OS of the k8s node is google [COS](https://cloud.google.com/kubernetes-engine/docs/concepts/node-images#cos).
- More detail about the k8s node:
```
    cloud.google.com/gke-os-distribution: cos
    kubernetes.io/arch: amd64
    kubernetes.io/os: linux
```
- The prometheus version `v2.22.1 (Nov 5, 2020)` runs as pod inside the GKE cluster.

**Comment by codesome** [comment]
I am confused why `process_resident_memory_bytes` and `container_memory_rss` are so different especially the former being higher since that is the process memory. Either way, both should not increase in memory so much and I don't see any place where we would load so much into memory at once for compaction. So I am suspecting this to be OS memory handling issue for the file memory-mapping for now. Investigating more.

**Comment by codesome** [comment]
Did not find anything concrete in Prometheus that would take up so much time, it depends on the OS/Go to take back the memory what we just stopped referencing in Prometheus.

**Comment by shay-berman** [comment]
Some updates @brian-brazil @SuperQ  @codesome 

**TLDR** It looks like that if you reduce the memory limit of prometheus instance then the same compaction process will just use less memory(up to its memory limit) and it will not impact the compaction run time or preformance.   So looks like prometheus uses more memory then it really needs, at least from the `process_resident_memory_bytes` metric prespective.

Its not clear why `process_resident_memory_bytes` going so high compare to `container_memory_working_set_bytes`, but based on the result here it looks like  `container_memory_working_set_bytes` metric indicate much better about the real consumption of prometheus.

**Still open questions:**
1. Why prometheus `process_resident_memory_bytes` metric going so high compare to `container_memory_working_set_bytes`. Maybe its related to the new prometheus feature of [memory mapping](https://grafana.com/blog/2020/06/10/new-in-prometheus-v2.19.0-memory-mapping-of-full-chunks-of-the-head-block-reduces-memory-usage-by-as-much-as-40/).
2. If its an expected behaivour from prometheus side then I would expect it to be documented somewhere. Probably other prometheus uses may hit the same issue and wonder about this memory consumption and understand that its better to monitor `container_memory_working_set_bytes` vs the `process_resident_memory_bytes`.


**More details about the TLDR**
I wanted to investigate it deeper, so I actually took 3 prometheus instances(4, 5 & 8) and set there resource.memory.limits from 21GB to 9GB while keeping the high resource.memory.limit of 21GB for all the rest of my prometheus instances (1, 2, 3, 6, 7, 9). And then I just waited for the next memory spike from `process_resident_memory_bytes` metric...

Eventually the spike arrived like before and it was triggered by prometheus compaction (same as previous spike) :
![image](https://user-images.githubusercontent.com/1079649/105058685-2602db80-5a7f-11eb-9128-50e14b4d7610.png)

As you can see in the screenshot the `process_resident_memory_bytes` metric spiked up to 18.4GB but only for the prometheus instances with the 21GB limit (again while the `container_memory_working_set_bytes` and `container_memory_rss` are not spiked in the same extent it keep below 5.8GB).   On the other hand, the prometheus instances (with the resource.memory.limit to 9GB) spiked as well but only below there memory limit. And it does not impact the compaction time as well(I checked the log and the compaction finished the same as all other prometheus compaction at the same spike).

I still don't know why `process_resident_memory_bytes` is going so high vs the `container_memory_working_set_bytes`.
It looks like that somehow prometheus or the kernel that runs it actually manage its memory and evict things that not needed when it reached its memory limit. Looks like if you add more memory promethes will just try and use it all even if not really needed.

**Comment by roidelapluie** [comment]
On Linux, the go runtime uses MADV_FREE to release unused memory. This is more efficient but may result in higher reported RSS. The kernel will reclaim the unused data when it is needed.

**Comment by SuperQ** [comment]
Yes, I agree, this is a side effect of the way Linux deals with accounting of RSS. Basically, RSS is a lie, and can't be trusted when using MADV_FREE.

Unfortunately, there's not a lot Prometheus can do about this. It's between Go and the Linux kernel to fix reporting of memory allocations.

You can inject `GODEBUG=madvdontneed=1` into the container to disable MADV_FREE, but I don't recommend this. This will eat more CPU time and slow down the host OS for other containers.

**Comment by codesome** [comment]
Looking at the Cortex mixins that we use to track the memory, I think `container_memory_working_set_bytes` and `go_memstats_heap_inuse_bytes` would give you a better picture of memory usage while the former is the one that will give you an idea of when Prometheus would be OOM killed.

**Comment by shay-berman** [comment]
Thanks a lot guys for your collaboration on this ticket.

So based on the analysis and the golang documentation @roidelapluie mentioned :
![image](https://user-images.githubusercontent.com/1079649/105358504-ea8f1b00-5bfe-11eb-9661-a68eb3288482.png)

**So to summarize this long discussion:**
1. The `process_resident_memory_bytes` metric that prometheus exposes do not really reflect the memory usage of 
 prometheus, this metric may indicate on a higher memory usage then the prometheus is actually using.
2. We are not sure 100% by but we think its related to Golang [implementation](https://golang.org/doc/go1.12#runtime) and OS kernel behavior of releasing memory. 
3. Its better to observe prometheus memory usage by `container_memory_working_set_bytes` metric (You have this metric only if you run it from k8s).
4. About `go_memstats_heap_inuse_bytes` metric its probably also good indication about prometheus memory usage, but its showing a bit lower numbers then `container_memory_working_set_bytes`. So not sure 100% if we can rely on it but it defiantly more align with `container_memory_working_set_bytes` metric then `process_resident_memory_bytes`.

@codesome @SuperQ @roidelapluie please approve and let me know if I can close this ticket.

**Comment by SuperQ** [comment]
It looks like [Go 1.16 will use MADV_DONTNEED](https://github.com/golang/go/issues/42894) like 1.12 and earlier did.

**Comment by MrYueQ** [comment]
Hi. I also encountered this kind of problem. It was found that the cache was not released. It's easy to get OOM from the kernel。
![image](https://user-images.githubusercontent.com/14846522/113077164-3237b180-9203-11eb-9b42-ab42fcf396c9.png)
